### ISSUE TITLE
Performance issue when calling Keras conv2D / gen_nn_ops.conv2d with different sizes

### ISSUE THREAD

**Issue by farotem** [tensorflow/tensorflow#41325]
### System information

-   **Have I written custom code (as opposed to using a stock example script
    provided in TensorFlow)**:
yes
-   **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:
Linux Ubuntu 18.04
-   **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue
NA
    happens on a mobile device**:
NA
-   **TensorFlow installed from (source or binary)**:  binary wheel via PyPI
-   **TensorFlow version (use command below)**:
v2.1.0-rc2-17-ge5bf8de 2.1.0;
v2.2.0-0-g2b96f3662b 2.2.0
-   **Python version**:
-   **Bazel version (if compiling from source)**:
-   **GCC/Compiler version (if compiling from source)**:
-   **CUDA/cuDNN version**:
CUDA 10.0
-   **GPU model and memory**:
V100 32 GB

### Describe the problem
I'm suspecting time performance issue when running a model with different input sizes.
the first time of running a model produce tracing, but I created a model with tf.function with tensor spec and yet when changing the input size (even by 1) the performance decreases a lot.
the model is build with keras layer conv2d.
when running the low level tensorflow API gen_nn_ops.conv2d received the same decrease of performance.

**Describe the expected behavior**
the timing performance shouldn't decrease after the 1st time of calling the model with different size.

Keras code:

````
import tensorflow as tf
import numpy as np
from time import perf_counter
from tensorflow.keras.layers import Conv2D

filters = 8
kernel_size = (3, 3)
inp_shape = (15, 15, filters)

layers = [Conv2D(filters=filters, kernel_size=kernel_size, input_shape=inp_shape)]
layers += [Conv2D(filters=filters, kernel_size=kernel_size) for _ in range(4)]
model = tf.keras.Sequential(layers=layers)

@tf.function(input_signature=[tf.TensorSpec([None, 15, 15, 8])])
def tf_func(inp):
    print('graph tracing')
    return model(inp)

def prepare_inp(n):
    return np.random.rand(n, *inp_shape).astype("float32")

out = tf_func(prepare_inp(1)).numpy() # first call - very slow
N = 30000
for n in range(N, N+6):
    inp = prepare_inp(n)
    elapsed = []
    for _ in range(5):
        _start = perf_counter()
        out = tf_func(inp).numpy()
        elapsed.append(perf_counter() - _start)
    print(f'[{n:,} samples] running time (msec): ' + ' '.join(f'{1000*e:6.1f}' for e in elapsed))

"""
Output:
graph tracing
[30,000 samples] running time (msec):  466.6   77.4   63.0   63.5   77.9
[30,001 samples] running time (msec):  389.0   64.0   49.6   38.8   63.4

Expected something like:
graph tracing
[30,000 samples] running time (msec):  466.6   77.4   63.0   63.5   77.9
[30,001 samples] running time (msec):   60.0   64.0   49.6   38.8   63.4
"""
````

low level API code:

```
import tensorflow as tf
import numpy as np
from time import perf_counter

filters = 8
kernel_size = (3, 3)
order = 'NHWC' # or 'NCHW'
img_shape = (15, 15)

inp_shape = img_shape + (filters,) if order == 'NHWC' else (filters,) + img_shape
kernel = tf.constant(np.random.rand(*kernel_size, filters, filters), dtype='float32')

from tensorflow.python.ops import gen_nn_ops

@tf.function(input_signature=[tf.TensorSpec([None, 15, 15, 8])])
def tf_func(inp):
    print('graph tracing')
    net = inp
    for _ in range(5):
        net = gen_nn_ops.conv2d(net, kernel, (1, 1, 1, 1), 'VALID', data_format=order)
    return net

def prepare_inp(n):
    return np.random.rand(n, *inp_shape).astype("float32")

out = tf_func(prepare_inp(1)).numpy() # first call - very slow
N = 30010
for n in range(N, N+6):
    inp = prepare_inp(n)
    elapsed = []
    for _ in range(5):
        _start = perf_counter()
        out = tf_func(inp).numpy()
        elapsed.append(perf_counter() - _start)
    print(f'[{n:,} samples] running time (msec): ' + ' '.join(f'{1000*e:6.1f}' for e in elapsed))

"""
Output:
graph tracing
[30,010 samples] running time (msec):  466.6   77.4   63.0   63.5   77.9
[30,011 samples] running time (msec):  389.0   64.0   49.6   38.8   63.4

Expected something like:
graph tracing
[30,010 samples] running time (msec):  466.6   77.4   63.0   63.5   77.9
[30,011 samples] running time (msec):   60.0   64.0   49.6   38.8   63.4
"""
```

**Comment by amahendrakar** [comment]
Was able to reproduce the issue, please find the gist of it [here](https://colab.research.google.com/gist/amahendrakar/c78ab0058bd22aa0f588e6528f8d1c96/41325.ipynb). Thanks!

**Comment by farotem** [comment]
I ran the gist, it looks like it runs with CPU in the gist but in my environment it runs with GPU.
the issue occurs only in GPU.

**Comment by amahendrakar** [comment]
> I ran the gist, it looks like it runs with CPU in the gist but in my environment it runs with GPU.
> the issue occurs only in GPU. what should I change so it will run in gist on the GPU?

@farotem,
Please click on "Runtime" -> "Change runtime type"  and then select the preferred hardware accelerator.

**Comment by farotem** [comment]
please add the label GPU since the issue is happening only on GPU

**Comment by rmothukuru** [comment]
@farotem,
Sorry for the delayed response. **`Performance`** seems to be improved when your code is executed with **`Tensorflow Version 2.5`**. Please find [the Gist](https://colab.research.google.com/gist/rmothukuru/742c2f26f882c0c82d2e398e1b362f18/41325.ipynb). Can you please confirm the same? Thanks!

**Comment by farotem** [comment]
@rmothukuru 
sorry for the late delay but it looks like it didn't improved with tensorflow 2.5.
the gist mechanism changed a bit.
when running the gist for the 1st time the performance issue occurs. 
when calling **run cell** again with the same inputs shapes it doesn't.
when calling run cell again with **different inputs shape** (N = 30010 for example) its happening again.
can you please re open it?

**Comment by sushreebarsa** [comment]
@farotem Can you please try with the latest TF v2.7.0 and let us know if it helps?Thanks!

**Comment by farotem** [comment]
Hi,
I checked on tf 2.7 and the issue is still happening 

tf version = 2.7.0
graph tracing
[29,990 samples] running time (msec): 1567.1  128.6  125.4  127.3  125.2
[29,991 samples] running time (msec): 1377.4  125.5  125.9  124.9  124.5
[29,992 samples] running time (msec): 1365.7  128.6  125.1  125.3  125.3
[29,993 samples] running time (msec): 1377.8  125.7  126.0  124.9  124.3
[29,994 samples] running time (msec): 1366.3  124.1  124.7  125.0  125.5
[29,995 samples] running time (msec): 1380.6  126.2  125.6  124.1  124.9

**Comment by gowthamkpr** [comment]
@farotem This is not a bug in tensorflow, everytime the input size changes, graph retracing happens as to accommodate the input size as mentioned [here](https://www.tensorflow.org/guide/function#controlling_retracing) and that is the reason why performance is being reduced. To control the tracing behavior, you can use the following techniques as mentioned [here](https://www.tensorflow.org/guide/function#controlling_retracing). Thanks!

**Comment by farotem** [comment]
@gowthamkpr unfortunately from my benchmark calling conv2d with tf functions and tensor spec did not help to improve the performance when changing the batch size (you can see the example I shared) do you have a suggestion what to change to make sure retracing is not accuring when using conv2d?

**Comment by gowthamkpr** [comment]
@farotem The issue here is that, if you call the layer with inputs of different sizes, each time you call a layer your Function retraces a new graph for every call, you'll find that your code executes more slowly than if you didn't use tf.function. I would recoomend you to use unknown dimensions for flexibility as mentioned [here](https://www.tensorflow.org/guide/function#use_unknown_dimensions_for_flexibility) to control retracing
